module2/projects/coupon-codes-api/index.md

```diff
@@ -0,0 +1,467 @@
+  * RSpec, Pry, Simplecov, Shoulda-Matchers, Orderly, jsonapi-serializer (or other serializer gem), HTTParty, Faker, Faraday, and FactoryBot are pre-approved
+* Scaffolding is not permitted on this project.
+
+## Setup
+
+This project is an extension of Little Shop. Students have two options for setup:
+
+1. If your Little Shop project is complete, you can use it as a starting point for this project. If you are not the repo owner, fork the project to your account. If you are the repo owner, you can work off the repo without forking, just make sure your teammates have a chance to fork before pushing any commits to your repo.
+1. If your Little Shop project is _not_ complete, fork ****Update** [this repo](need to update)** as a starting point for this project.
+
+## Evaluation
+Evaluation information for this project can be found [here](./evaluation).
+
+-----
+
+## Functionality Overview
+
+* A Coupon belongs to a Merchant
+* An Invoice optionally belongs to a Coupon. An invoice may only have one coupon.
+  * Note: When creating this new association on Invoice, your existing tests will fail unless the association is optional. Use these guides as a reference.
+  * You are not required to build functionality for a user applying a coupon to an invoice, but can instead use test data, Rails console or seed data to add coupons to existing invoices to verify behavior.
+* You should be full CRUD functionality for coupons with criteria/restrictions defined below:
+   - A merchant can have a maximum of 5 activated coupons in the system at one time.
+   - A merchant cannot delete a coupon, rather they can activate/deactivate them.
+   - A Coupon has a name, unique code (e.g. "BOGO50"), and either percent-off or dollar-off value. The coupon's code must be unique in the whole database.
+* If a coupon's dollar value (ex. "$10 off") exceeds the total cost of that merchant's items on the invoice, the grand total for that merchant's items should then be $0. (In other words, the merchant will never *owe* money to a customer.)
+* A coupon code from a Merchant only applies to Items sold by that Merchant.
+
+​
+## 1) CRUD Endpoints
+
+Below is the expected JSON response for each request. We have also outlined a few examples of Sad Paths you may consider adding in. In your project, you should take time to implement at least 2 sad paths total, but you are not limited to the examples we provide. 
+
+<section class="dropdown">
+### 1. Merchant Coupon Show Page 
+
+Returns a specific coupon and shows a count of how many times that coupon has been used.
+
+**Request**
+```bash
+GET /api/v1/merchants/:merchant_id/coupons/:id
+Content-Type: application/json
+Accept: application/json
+```
+
+**Response**
+
+```json
+status: 200
+body:
+
+{
+  "data": {
+    "id": "1",
+    "type": "coupon",
+    "attributes": {
+      "name": "Buy One Get One 50",
+      "code": "BOGO50",
+      "amount_off": 50,
+      "percent_off": null,
```

Comment:
I'd be curious what @abdulredd thinks of this schema cause he's run this existing coupon code project more than me. I've seen students use a `type` attribute (percent or dollar) and then an `amount` attribute which is the integer amount. So `type` of percent and `amount` of 20 would be 20% off, for example. This would avoid nullable fields. I think, by convention, if a field is null, it's also just not included in a JSON response. So we could try that out if we keep the schema like this. 

Part of the interesting part of the project is students working through how to design this in the database, so I'm wondering if we should have some flexibility in how the attributes are displayed? I guess we can't if we require it to work on the FE though...

Reply:
Yeah I also would probably do a type and then amount. But I didn't know if that felt too complicated, but it's probably not. Like you said, it gives them a chance to make and talk about design decisions. Perhaps one of the FE tasks can be adjusting the FE code to add and/or properly center the coupon value?

Reply:
Here's what I'm thinking for this after a little more time mulling it over.

The FE portion of this project is students adding coupon info to their existing merchant pages (or whatever makes sense with what they were given for the group project). This way, they also have to design the schema for the Coupon table which is an important skill for us to assess.

We can either remove example responses altogether, or I can edit them to look something like this:
```bash
status: 200
body:

{
  "data": {
    "id": "1",
    "type": "coupon",
    "attributes": {
      "name": "Buy One Get One 50",
      ...remaining attributes go here...
    }
  }
}
```

Thoughts?